Fix PSScriptAnalyzer warnings

bergmeister opened the pull request:
To help fixing #3791 
Fixes a few hundred warnings of type PSPossibleIncorrectComparisonWithNull, PSAvoidUsingCmdletAliases (mostly for ForEach-Object, Where-Object, Select-Object) and some PSUseDeclaredVarsMoreThanAssignment warnings.
As part of this, it was also found that a test in ParameterBindingTests needed improvement because since the null comparison is now correct, the test failed initially.

bergmeister committed:
Fixed most PSAvoidUsingCmdletAliases warnings for ForEach-Object (%), Where-Object (?), Select-Object and a few others.
Fixed some PSUseDeclaredVarsMoreThanAssignment warnings (one must be careful with them as some of those warnings are false positives)

bergmeister committed:
Fixed a problem  in ParameterBinding.Tests.ps1 test because the  null comparison is now correct and also removed unnecessary initialisation that was not needed for the test but made it difficult to know what was going. Enhanced the test.
Fixed one typo in Dynamicparameters.Tests.ps1 that got introduced during the last refactoring and made 2 tests fail.

msftclas commented:
@bergmeister,
Thanks for having already signed the Contribution License Agreement. Your agreement has not been validated yet. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

iSazonov commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -47,8 +47,8 @@ function Get-CodeCoverageChange($r1, $r2, [string[]]$ClassName)
-            $c1 = $r1.Assembly.ClassCoverage | ?{$_.ClassName -eq $Class }
-            $c2 = $r2.Assembly.ClassCoverage | ?{$_.ClassName -eq $Class }
+            $c1 = $r1.Assembly.ClassCoverage | Where-Object{$_.ClassName -eq $Class }
+            $c2 = $r2.Assembly.ClassCoverage | Where-Object{$_.ClassName -eq $Class }
```
Please add space after `Where-Object` in both lines.

iSazonov commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -136,7 +136,7 @@ function Get-CoverageData($xmlPath)
     }
     $CoverageData.PSTypeNames.Insert(0,"OpenCover.CoverageData")
-    Add-Member -InputObject $CoverageData -MemberType ScriptMethod -Name GetClassCoverage -Value { param ( $name ) $this.assembly.classcoverage | ?{$_.classname -match $name } }
+    Add-Member -InputObject $CoverageData -MemberType ScriptMethod -Name GetClassCoverage -Value { param ( $name ) $this.assembly.classcoverage | Where-Object{$_.classname -match $name } }
```
Please add space after `Where-Object`.

iSazonov reviewed (commented):
As requested in #3791 we should add meta test with scriptanalyzer.

bergmeister committed:
Refactoring fixes as suggested at the pull request review. Mainly includes spaces after Where-Object/Select-Object, 4 lines being reverted due to accidental changes and minor test refactoring.
The test refactoring idea was basically to change '$null -eq $something | Should Be $true' to '$something | Should BeNullOrEmpty'

bergmeister commented:
Thanks @iSazonov for taking such a detailed look and spotting the accidental mistakes that I made when going through the hundreds of refactorings. I have followed all your suggestions and submitted the fix as you can see above. The total number of warnings seems to have gone down by about 350.

iSazonov commented on src/System.Management.Automation/engine/Types_Ps1Xml.cs:
```diff
@@ -232,10 +232,10 @@ public static IEnumerable<TypeData> Get()
             td24.Members.Add("IPV4Address",
                 new ScriptPropertyData(@"IPV4Address", GetScriptBlock(@"$iphost = [System.Net.Dns]::GetHostEntry($this.address)
-          $iphost.AddressList | Where-Object{ $_.AddressFamily -eq [System.Net.Sockets.AddressFamily]::InterNetwork } | Select-Object -first 1"), null));
+          $iphost.AddressList | Where-Object{  $_.AddressFamily -eq [System.Net.Sockets.AddressFamily]::InterNetwork } | Select-Object -first 1"), null));
```
Please remove extra space before `$_` and add space after `Where-Object` and before `{`.

The same typo is below.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Security/AclCmdlets.Tests.ps1:
```diff
@@ -12,7 +12,7 @@ Describe "Acl cmdlets are available and operate properly" -Tag CI {
 
         $newacl = get-acl $directory
-        $newrule = $newacl.Access | Where-Object{ $accessrule.FileSystemRights -eq $_.FileSystemRights -and $accessrule.AccessControlType -eq $_.AccessControlType -and $accessrule.IdentityReference -eq $_.IdentityReference }
+        $newrule = $newacl.Access | Where-Object{  $accessrule.FileSystemRights -eq $_.FileSystemRights -and $accessrule.AccessControlType -eq $_.AccessControlType -and $accessrule.IdentityReference -eq $_.IdentityReference }
```
Please add space after `Where-Object` and remove extra space before `$accessrule`.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -808,7 +808,7 @@ Describe "Import-PSSession functional tests" -tags "Feature" {
 
         It "Temporary module should be automatically removed after runspace is closed" -Skip:$skipTest {
-            ($null -eq (Get-Module | Where-Object { $_.Path -eq $module.Path })) | Should Be $true
+             (Get-Module | Where-Object { $_.Path -eq $module.Path }) | Should BeNullOrEmpty
```
Indentation.

iSazonov commented on test/powershell/Modules/PackageManagement/PackageManagement.Tests.ps1:
```diff
@@ -32,9 +32,9 @@ Describe "PackageManagement Acceptance Test" -Tags "Feature" {
         $gpp = Get-PackageProvider
 
-        $gpp | Where-Object{ $_.name -eq "NuGet" } | should not BeNullOrEmpty
+        $gpp | Where-Object{  $_.name -eq "NuGet" } | should not BeNullOrEmpty
```
Please remove extra space.

iSazonov commented on test/powershell/Modules/PackageManagement/PackageManagement.Tests.ps1:
```diff
@@ -32,9 +32,9 @@ Describe "PackageManagement Acceptance Test" -Tags "Feature" {
+        $gpp | Where-Object{  $_.name -eq "NuGet" } | should not BeNullOrEmpty
 
-        $gpp | Where-Object{ $_.name -eq "PowerShellGet" } | should not BeNullOrEmpty
+        $gpp | Where-Object{  $_.name -eq "PowerShellGet" } | should not BeNullOrEmpty
```
Please remove extra space.

iSazonov commented on test/tools/OpenCover/OpenCover.psm1:
```diff
@@ -136,7 +136,7 @@ function Get-CoverageData($xmlPath)
     }
     $CoverageData.PSTypeNames.Insert(0,"OpenCover.CoverageData")
-    Add-Member -InputObject $CoverageData -MemberType ScriptMethod -Name GetClassCoverage -Value { param ( $name ) $this.assembly.classcoverage | Where-Object{$_.classname -match $name } }
+    Add-Member -InputObject $CoverageData -MemberType ScriptMethod -Name GetClassCoverage -Value { param ( $name ) $this.assembly.classcoverage | Where-Object{ $_.classname -match $name } }
```
Please remove extra space and add space after `Where-Object`.

lzybkr commented on src/Modules/Windows-Full/Microsoft.PowerShell.ODataUtils/Microsoft.PowerShell.ODataV4Adapter.ps1:
```diff
@@ -1495,7 +1495,7 @@ function SaveCDXMLSingletonCmdlets
 		$xmlWriter = SaveCDXMLHeader $xmlWriter $Uri $singletonName $singletonName $CmdletAdapter
 
-        if ($null -eq $associatedEntityType.BaseTyp -and $null -ne $associatedEntityType.BaseTypeStr -and $associatedEntityType.BaseTypeStr -ne '')
+        if ($null -eq $associatedEntityType.BaseType -and $null -ne $associatedEntityType.BaseTypeStr -and $associatedEntityType.BaseTypeStr -ne '')
```
@anmenaga - this looks like a bug that should get fixed in Windows PowerShell.

bergmeister replied:
@lzybkr : I'm not sure what you mean but the missing 'e' on 'BaseType' was an accidental deletion that happened in the first commit that I corrected

lzybkr replied:
I see - this is a good example of why I'm not a fan of mixing minor changes in one big commit, because I do sometimes review commits by themselves instead of the full PR.

lzybkr commented:
I'm seeing different sorts of changes in a single commit - and with this many changes, it would be easier and safer to keep each refactoring as a single commit.

iSazonov commented:
+1 for multiple commits. Maybe:
1. Build.psm1
2. Rest tool scripts
3. Test scripts
4. Rest files.

bergmeister committed:
Fix spacing around Where-Object and indentation as requested in the second pull request review

bergmeister commented:
@iSazonov I have fixed your requested changes about the fine-tuning of the spacing. When I started out doing that, I initially wanted to fix only most PSPossibleIncorrectComparisonWithNull violations but then I came across lots of other things to be improved like e.g. the aliases. Because the contribution guide [here](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#lifecycle-of-a-pull-request) recommended to have just one commit per change, I thought I'd try to put it all into one commit for the pull request. I am not sure how one would change the  commit history now that the commits are already pushed to the remote. Sorry, maybe I should have rather checked with you before.

iSazonov commented:
You would rebase and split - see https://stackoverflow.com/questions/6217156/break-a-previous-commit-into-multiple-commits

bergmeister commented:
@iSazonov OK, thanks. I did not feel confident/adventurous enough to rebase commits that are already pushed to the remote, so I have created a new branch, pulled in my commit changes and made then the 4 separate commits as requested by you. I have created the new PR #4213 for that, so feel free to close this one.
